### PR TITLE
[KOGITO-2391] introducing ReteEvaluator

### PR DESCRIPTION
**Pull request by mariofusco**
**Thank you for submitting this pull request**

**NOTE!:** Double check the target branch for this PR.
The default is `main` so it will target Drools 8 / Kogito / Optaplanner 8.
If this PR is not strictly related to drools and kogito project in `drools.git`, it should probably target `7.x`as a branch

**Ports** If a forward-port or a backport is needed, paste the forward port PR here

[link](https://www.example.com)

**JIRA**: https://issues.redhat.com/browse/KOGITO-2391

Related pull request:
https://github.com/kiegroup/drools/pull/3961
https://github.com/kiegroup/kogito-runtimes/pull/1688
https://github.com/kiegroup/optaplanner/pull/1643

### CODE REVIEW

File: drools-core/src/main/java/org/drools/core/base/DefaultKnowledgeHelper.java [hunk 1]
```diff
@@ -83,7 +84,7 @@
 
     protected Activation                                activation;
     protected Tuple                                     tuple;
-    protected WrappedStatefulKnowledgeSessionForRHS   workingMemory;
+    protected ReteEvaluatorForRHS workingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/base/DefaultKnowledgeHelper.java]: Rename workingMemory as reteEvaluator

File: drools-core/src/main/java/org/drools/core/base/ReteEvaluatorForRHS.java [hunk 2]
```diff
@@ -79,7 +79,7 @@
 /**
  * Wrapper of StatefulKnowledgeSessionImpl so to intercept call from RHS internal Drools execution and proxy or delegate method call as appropriate.
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/base/ReteEvaluatorForRHS.java]: Javadoc to be updated

File: drools-core/src/main/java/org/drools/core/common/ReteEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2010 Red Hat, Inc. and/or its affiliates.
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/common/ReteEvaluator.java]:
```suggestion
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
```

File: drools-core/src/main/java/org/drools/core/impl/StatefulKnowledgeSessionImpl.java [hunk 4]
```diff
@@ -1987,6 +1970,11 @@ public void endOperation() {
         }
     }
 
+    @Override
+    public void executeQueuedActions() {
+
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/impl/StatefulKnowledgeSessionImpl.java]: Empty method?

File: drools-core/src/main/java/org/drools/core/reteoo/AccumulateNode.java [hunk 5]
```diff
@@ -135,18 +135,17 @@ public BetaConstraints getResultBinder() {
...
 
     public InternalFactHandle createResultFactHandle(final PropagationContext context,
-                                                     final InternalWorkingMemory workingMemory,
+                                                     final ReteEvaluator reteEvaluator,
                                                      final LeftTuple leftTuple,
                                                      final Object result) {
         InternalFactHandle handle = null;
         if ( context.getReaderContext() != null ) {
-            handle = context.getReaderContext().createAccumulateHandle( context.getEntryPoint(), workingMemory, leftTuple, result, getId() );
+            handle = context.getReaderContext().createAccumulateHandle( context.getEntryPoint(), (InternalWorkingMemory) reteEvaluator, leftTuple, result, getId() );
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/AccumulateNode.java]: This cast breaks the abstraction

File: drools-core/src/main/java/org/drools/core/reteoo/CompositePartitionAwareObjectSinkAdapter.java [hunk 6]
```diff
@@ -106,8 +106,8 @@ public void changeSinkPartition( ObjectSink sink, RuleBasePartitionId oldPartiti
     }
 
     @Override
-    public void propagateAssertObject( InternalFactHandle factHandle, PropagationContext context, InternalWorkingMemory workingMemory ) {
-        CompositeDefaultAgenda compositeAgenda = (CompositeDefaultAgenda) workingMemory.getAgenda();
+    public void propagateAssertObject( InternalFactHandle factHandle, PropagationContext context, ReteEvaluator reteEvaluator ) {
+        CompositeDefaultAgenda compositeAgenda = (CompositeDefaultAgenda) reteEvaluator.getActivationsManager();
         if (hashed) {
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/CompositePartitionAwareObjectSinkAdapter.java:111]: Is activationsManager also a CompositeDefaultAgenda?

File: drools-core/src/main/java/org/drools/core/reteoo/ObjectTypeNode.java [hunk 7]
```diff
@@ -536,7 +528,7 @@ public void mergeExpirationOffset(ObjectTypeNode other) {
         @Override
         public void execute(JobContext ctx) {
             ExpireJobContext context = (ExpireJobContext) ctx;
-            context.workingMemory.queueWorkingMemoryAction(context.expireAction);
+            context.reteEvaluator.addPropagation(context.expireAction, true);
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/ObjectTypeNode.java:531]: Is this equivalent?

File: drools-core/src/main/java/org/drools/core/reteoo/QueryElementNode.java [hunk 8]
```diff
@@ -122,15 +119,15 @@ public boolean isOpenQuery() {
 
     @SuppressWarnings("unchecked")
     public InternalFactHandle createFactHandle(final PropagationContext context,
-                                               final InternalWorkingMemory workingMemory,
+                                               final ReteEvaluator reteEvaluator,
                                                final LeftTuple leftTuple ) {
         InternalFactHandle handle = null;
         if( context.getReaderContext() != null ) {
-            handle = context.getReaderContext().createQueryHandle( leftTuple, workingMemory, getId() );
+            handle = context.getReaderContext().createQueryHandle( leftTuple, (InternalWorkingMemory) reteEvaluator, getId() );
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/QueryElementNode.java]: This cast breaks the abstraction

File: drools-core/src/main/java/org/drools/core/reteoo/QueryElementNode.java [hunk 9]
```diff
@@ -278,17 +273,18 @@ public void rowAdded(final RuleImpl rule,
...
-                                                                          workingMemory,
+                                                                          reteEvaluator,
                                                                           objects);
             
             RightTuple rightTuple = createResultRightTuple(resultHandle, resultLeftTuple, dquery.isOpen());
 
             boolean pass = true;
-            if ( query.isAbductive() ) {
+            if ( reteEvaluator instanceof InternalWorkingMemory && query.isAbductive() ) {
+                InternalWorkingMemory workingMemory = (InternalWorkingMemory) reteEvaluator;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/QueryElementNode.java:287]: This breaks the abstraction.
Moreover what if this if fails?

File: drools-core/src/main/java/org/drools/core/reteoo/Rete.java [hunk 10]
```diff
@@ -16,20 +16,15 @@
...
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
-import org.drools.core.WorkingMemoryEntryPoint;
-import org.drools.core.common.DroolsObjectInputStream;
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
+import org.drools.core.common.ReteEvaluator;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/Rete.java:27]: Is it correct to use both ReteEvaluator and InternalWorkingMemory in the same class?

File: drools-core/src/main/java/org/drools/core/reteoo/ReteooFactHandleFactory.java [hunk 11]
```diff
@@ -48,38 +48,43 @@ public InternalFactHandle newFactHandle( final long id,
...
             return new DefaultFactHandle( id,
                                           object,
                                           recency,
-                                          wmEntryPoint != null ? wmEntryPoint : workingMemory,
+                                          getWmEntryPoint(reteEvaluator, wmEntryPoint),
                                           conf != null && conf.isTrait() );
         }
     }
 
+    private WorkingMemoryEntryPoint getWmEntryPoint(ReteEvaluator reteEvaluator, WorkingMemoryEntryPoint wmEntryPoint) {
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/ReteooFactHandleFactory.java]: Why this change?

File: drools-core/src/main/java/org/drools/core/reteoo/RightInputAdapterNode.java [hunk 12]
```diff
@@ -326,7 +327,7 @@ public BitMask getLeftInferredMask() {
     }
 
     @Override
-    public void updateSink(ObjectSink sink, PropagationContext context, InternalWorkingMemory workingMemory) {
+    public void updateSink(ObjectSink sink, PropagationContext context, InternalWorkingMemory wm) {
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/RightInputAdapterNode.java:330]: This method still uses InternalWorkingMemory

File: drools-core/src/main/java/org/drools/core/reteoo/RightTuple.java [hunk 13]
```diff
@@ -18,6 +18,7 @@
 
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
+import org.drools.core.common.ReteEvaluator;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/RightTuple.java:21]: Is it possible to remove usage of InternalWorkingMemory?

File: drools-core/src/main/java/org/drools/core/reteoo/WindowNode.java [hunk 14]
```diff
@@ -27,6 +27,7 @@
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/reteoo/WindowNode.java:27]: Is it possible to remove InternalWorkingMemory usage?

File: drools-core/src/main/java/org/drools/core/rule/Declaration.java [hunk 15]
```diff
@@ -26,6 +26,7 @@
 import org.drools.core.common.DroolsObjectInputStream;
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/rule/Declaration.java:28]: Is it possible to remove it?

File: drools-core/src/main/java/org/drools/core/rule/EvalCondition.java [hunk 16]
```diff
@@ -26,6 +26,7 @@
 import java.util.Map;
 
 import org.drools.core.WorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/rule/EvalCondition.java:28]: Is it possible to remove it?

File: drools-core/src/main/java/org/drools/core/rule/PredicateConstraint.java [hunk 17]
```diff
@@ -27,6 +27,7 @@
 
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/rule/PredicateConstraint.java:29]: Is it possible to remove it?

File: drools-core/src/main/java/org/drools/core/rule/ReturnValueRestriction.java [hunk 18]
```diff
@@ -27,6 +27,7 @@
 import org.drools.core.WorkingMemory;
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/rule/ReturnValueRestriction.java:29]: Is it possible to remove it?

File: drools-core/src/main/java/org/drools/core/rule/constraint/QueryNameConstraint.java [hunk 19]
```diff
@@ -20,6 +20,7 @@
 import org.drools.core.base.field.ObjectFieldImpl;
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/rule/constraint/QueryNameConstraint.java:22]: Is it possible to remove it?

File: drools-core/src/main/java/org/drools/core/runtime/rule/impl/OpenQueryViewChangedEventListenerAdapter.java [hunk 20]
```diff
@@ -20,6 +20,7 @@
 
 import org.drools.core.base.InternalViewChangedEventListener;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/runtime/rule/impl/OpenQueryViewChangedEventListenerAdapter.java:22]: Is it possible to remove it?

File: drools-core/src/main/java/org/drools/core/time/JobContext.java [hunk 21]
```diff
@@ -21,18 +21,19 @@
 
 import org.drools.core.common.InternalKnowledgeRuntime;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/time/JobContext.java:23]: Can be removed?

File: drools-core/src/main/java/org/drools/core/time/impl/CompositeMaxDurationTimer.java [hunk 22]
```diff
@@ -25,6 +25,7 @@
 import java.util.Map;
 
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/time/impl/CompositeMaxDurationTimer.java:27]: Can be removed?

File: drools-core/src/main/java/org/drools/core/time/impl/CronTimer.java [hunk 23]
```diff
@@ -24,6 +24,7 @@
 import java.util.Map;
 
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/time/impl/CronTimer.java:26]: Can be removed?

File: drools-core/src/main/java/org/drools/core/time/impl/DurationTimer.java [hunk 24]
```diff
@@ -27,6 +27,7 @@
 
 import org.drools.core.common.EventFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/time/impl/DurationTimer.java:29]: Can be removed?

File: drools-core/src/main/java/org/drools/core/time/impl/ExpressionIntervalTimer.java [hunk 25]
```diff
@@ -24,6 +24,7 @@
 import java.util.Map;
 
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/time/impl/ExpressionIntervalTimer.java:26]: Can be removed?

File: drools-core/src/main/java/org/drools/core/time/impl/IntervalTimer.java [hunk 26]
```diff
@@ -24,6 +24,7 @@
 import java.util.Map;
 
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/time/impl/IntervalTimer.java:26]: Can be removed?

File: drools-core/src/test/java/org/drools/core/reteoo/CrossProductTest.java [hunk 27]
```diff
@@ -23,6 +23,7 @@
 import java.util.List;
 
 import org.drools.core.WorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/test/java/org/drools/core/reteoo/CrossProductTest.java:25]: Can be removed?

File: drools-core/src/test/java/org/drools/core/reteoo/MockLeftTupleSink.java [hunk 28]
```diff
@@ -24,6 +24,7 @@
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/test/java/org/drools/core/reteoo/MockLeftTupleSink.java:24]: Can be removed?

File: drools-core/src/test/java/org/drools/core/reteoo/ReteooRuleBaseMultiThreadedTest.java [hunk 29]
```diff
@@ -18,6 +18,7 @@
 
 import org.drools.core.WorkingMemory;
 import org.drools.core.base.ClassFieldAccessorCache;
+import org.drools.core.common.ReteEvaluator;
 import org.drools.core.definitions.InternalKnowledgePackage;
```
Comment (danielezonca) [inline drools-core/src/test/java/org/drools/core/reteoo/ReteooRuleBaseMultiThreadedTest.java:22]: Can be removed?

File: drools-core/src/test/java/org/drools/core/reteoo/ReteooWorkingMemoryTest.java [hunk 30]
```diff
@@ -24,6 +24,7 @@
 import org.drools.core.common.EqualityKey;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/test/java/org/drools/core/reteoo/ReteooWorkingMemoryTest.java:25]: Can be removed?

File: drools-core/src/test/java/org/drools/core/time/impl/JDKTimerServiceTest.java [hunk 31]
```diff
@@ -16,9 +16,21 @@
...
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Stack;
+
 import org.drools.core.ClockType;
 import org.drools.core.SessionConfiguration;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-core/src/test/java/org/drools/core/time/impl/JDKTimerServiceTest.java:32]: Can be removed?

File: drools-core/src/test/java/org/drools/core/reteoo/builder/ReteooRuleBuilderTest.java [hunk 32]
```diff
@@ -69,7 +70,7 @@ public void testAddRuleWithPatterns() {
 
         final Consequence consequence = new Consequence() {
             public void evaluate(KnowledgeHelper knowledgeHelper,
-                                 WorkingMemory workingMemory) throws Exception {
```
Comment (danielezonca) [inline drools-core/src/test/java/org/drools/core/reteoo/builder/ReteooRuleBuilderTest.java:72]: Ca be removed?

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/LambdaConsequence.java [hunk 33]
```diff
@@ -24,6 +24,7 @@
 import org.drools.core.common.EventFactHandle;
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/LambdaConsequence.java:26]: Can be removed?

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/CombinedConstraint.java [hunk 34]
```diff
@@ -22,6 +22,7 @@
 import org.drools.core.RuleBaseConfiguration;
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/CombinedConstraint.java:24]: Can be removed?

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/ConstraintEvaluator.java [hunk 35]
```diff
@@ -22,6 +22,7 @@
 
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/ConstraintEvaluator.java:24]: Can be removed?

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/LambdaConstraint.java [hunk 36]
```diff
@@ -27,6 +27,7 @@
 import org.drools.core.base.field.ObjectFieldImpl;
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/LambdaConstraint.java:29]: Can be removed?

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/LambdaDataProvider.java [hunk 37]
```diff
@@ -22,6 +22,7 @@
 import java.util.Objects;
 
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/LambdaDataProvider.java:24]: Can be removed?

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/SupplierDataProvider.java [hunk 38]
```diff
@@ -21,6 +21,7 @@
 import java.util.Iterator;
 
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/constraints/SupplierDataProvider.java:23]: Can be removed?

File: drools-mvel/src/main/java/org/drools/mvel/MVELConstraint.java [hunk 39]
```diff
@@ -38,6 +38,7 @@
 import org.drools.core.common.DroolsObjectInputStream;
 import org.drools.core.common.InternalFactHandle;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-mvel/src/main/java/org/drools/mvel/MVELConstraint.java:40]: Can be removed?

File: drools-mvel/src/main/java/org/drools/mvel/MVELConstraint.java [hunk 40]
```diff
@@ -258,82 +259,82 @@ public boolean isAllowedCachedRight(Tuple tuple, ContextEntry context) {
...
         }
     }
 
-    protected ConditionEvaluator forceJitEvaluator(InternalFactHandle handle, InternalWorkingMemory workingMemory, Tuple tuple) {
-        ConditionEvaluator mvelEvaluator = createMvelConditionEvaluator(workingMemory);
+    protected ConditionEvaluator forceJitEvaluator(InternalFactHandle handle, ReteEvaluator reteEvaluator, Tuple tuple) {
+        ConditionEvaluator mvelEvaluator = createMvelConditionEvaluator(reteEvaluator);
         try {
-            mvelEvaluator.evaluate(handle, workingMemory, tuple);
+            mvelEvaluator.evaluate(handle, (InternalWorkingMemory) reteEvaluator, tuple);
```
Comment (danielezonca) [inline drools-mvel/src/main/java/org/drools/mvel/MVELConstraint.java:307]: This cast breaks the abstraction

File: drools-mvel/src/main/java/org/drools/mvel/MVELConstraint.java [hunk 41]
```diff
@@ -342,15 +343,15 @@ public void run() {
...
+    private ConditionEvaluator executeJitting(InternalFactHandle handle, ReteEvaluator reteEvaluator, Tuple tuple, ConditionEvaluator mvelEvaluator) {
+        InternalKnowledgeBase kBase = reteEvaluator.getKnowledgeBase();
         if (!isJmxAvailable() && MemoryUtil.permGenStats.isUsageThresholdExceeded(kBase.getConfiguration().getPermGenThreshold())) {
             return mvelEvaluator;
         }
 
         try {
             if (analyzedCondition == null) {
-                analyzedCondition = (( MVELConditionEvaluator ) mvelEvaluator).getAnalyzedCondition(handle, workingMemory, tuple);
+                analyzedCondition = (( MVELConditionEvaluator ) mvelEvaluator).getAnalyzedCondition(handle, (InternalWorkingMemory) reteEvaluator, tuple);
```
Comment (danielezonca) [inline drools-mvel/src/main/java/org/drools/mvel/MVELConstraint.java:354]: Same here

File: drools-mvel/src/main/java/org/drools/mvel/MVELConstraintBuilder.java [hunk 42]
```diff
@@ -50,6 +50,7 @@
 import org.drools.core.base.evaluators.EvaluatorDefinition;
 import org.drools.core.base.evaluators.Operator;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-mvel/src/main/java/org/drools/mvel/MVELConstraintBuilder.java:52]: Can be removed?

File: drools-mvel/src/main/java/org/drools/mvel/asm/ClassFieldAccessorFactory.java [hunk 43]
```diff
@@ -55,6 +55,7 @@
 import org.drools.core.base.extractors.BaseZonedDateTimeClassFieldReader;
 import org.drools.core.base.extractors.SelfReferenceClassFieldReader;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-mvel/src/main/java/org/drools/mvel/asm/ClassFieldAccessorFactory.java:57]: Can be removed?

File: drools-mvel/src/main/java/org/drools/mvel/expr/MVELConsequence.java [hunk 44]
```diff
@@ -23,6 +23,7 @@
 import org.drools.core.WorkingMemory;
 import org.drools.core.common.AgendaItem;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-mvel/src/main/java/org/drools/mvel/expr/MVELConsequence.java:25]: Can be removed?

File: drools-mvel/src/main/java/org/drools/mvel/expr/MVELEnabledExpression.java [hunk 45]
```diff
@@ -23,6 +23,7 @@
 
 import org.drools.core.WorkingMemory;
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-mvel/src/main/java/org/drools/mvel/expr/MVELEnabledExpression.java:25]: Can be removed?

File: drools-mvel/src/main/java/org/drools/mvel/expr/MVELObjectExpression.java [hunk 46]
```diff
@@ -20,6 +20,7 @@
 import java.io.ObjectOutput;
 
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-mvel/src/main/java/org/drools/mvel/expr/MVELObjectExpression.java:22]: Can be removed?

File: drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/AgendaEventSupportTest.java [hunk 47]
```diff
@@ -28,6 +28,7 @@
 import org.drools.core.base.evaluators.EvaluatorRegistry;
 import org.drools.core.common.InternalAgenda;
 import org.drools.core.common.InternalFactHandle;
+import org.drools.core.common.ReteEvaluator;
 import org.drools.core.definitions.InternalKnowledgePackage;
```
Comment (danielezonca) [inline drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/AgendaEventSupportTest.java:32]: Can be removed?

File: drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/KnowledgeBaseEventSupportTest.java [hunk 48]
```diff
@@ -25,6 +25,7 @@
 import org.drools.core.base.ClassFieldAccessorCache;
 import org.drools.core.base.ClassFieldAccessorStore;
 import org.drools.core.base.ClassObjectType;
+import org.drools.core.common.ReteEvaluator;
 import org.drools.core.definitions.InternalKnowledgePackage;
```
Comment (danielezonca) [inline drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/KnowledgeBaseEventSupportTest.java:29]: Can be removed?

File: drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/RuleBaseEventSupportTest.java [hunk 49]
```diff
@@ -111,7 +112,7 @@ public void setUp() throws Exception {
             private static final long serialVersionUID = 510l;
 
             public void evaluate(final KnowledgeHelper knowledgeHelper,
-                                 final WorkingMemory workingMemory) throws Exception {
```
Comment (danielezonca) [inline drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/RuleBaseEventSupportTest.java:114]: Can be removed?

File: drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/compiler/reteoo/MockLeftTupleSink.java [hunk 50]
```diff
@@ -25,6 +25,7 @@
 import org.drools.core.common.InternalWorkingMemory;
```
Comment (danielezonca) [inline drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/compiler/reteoo/MockLeftTupleSink.java:25]: Can be removed?

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/DroolsImpl.java [hunk 51]
```diff
@@ -57,7 +56,7 @@ public void insert(Object object) {
     @Override
     public void insert(Object object, boolean dynamic) {
         TerminalNode terminalNode = (( AgendaItem )getMatch()).getTerminalNode();
-        ((InternalWorkingMemoryActions)workingMemory).insert(object, dynamic, getRule(), terminalNode);
+        ((InternalWorkingMemoryEntryPoint)reteEvaluator.getDefaultEntryPoint()).insert(object, dynamic, getRule(), terminalNode);
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/DroolsImpl.java:59]: Is this cast acceptable?

File: drools-core/src/main/java/org/drools/core/impl/StatefulKnowledgeSessionImpl.java [hunk 52]
```diff
@@ -1779,23 +1762,23 @@ public PartitionAwareWorkingMemoryReteExpireAction(EventFactHandle factHandle, O
...
-        public void execute( InternalWorkingMemory wm ) {
+        public void execute( ReteEvaluator reteEvaluator ) {
             if (!factHandle.isValid()) {
                 return;
             }
 
-            PropagationContext context = createPropagationContextForFact( wm, factHandle, PropagationContext.Type.EXPIRATION );
-            ( (CompositeDefaultAgenda) wm.getAgenda() ).getPartitionedAgenda( partition ).registerExpiration( context );
+            PropagationContext context = createPropagationContextForFact( reteEvaluator, factHandle, PropagationContext.Type.EXPIRATION );
+            ( (CompositeDefaultAgenda) reteEvaluator.getActivationsManager() ).getPartitionedAgenda( partition ).registerExpiration( context );
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/impl/StatefulKnowledgeSessionImpl.java:1771]: Is this cast acceptable?

File: drools-core/src/main/java/org/drools/core/impl/StatefulKnowledgeSessionImpl.java [hunk 53]
```diff
@@ -2019,8 +2007,18 @@ public long getTimeToNextJob() {
         return this.timerService.getTimeToNextJob();
     }
 
-    public void addPropagation(PropagationEntry propagationEntry) {
-        agenda.addPropagation( propagationEntry );
+    @Override
+    public void addPropagation(PropagationEntry propagationEntry, boolean register) {
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/impl/StatefulKnowledgeSessionImpl.java:2006]: Is this refactor safe?

File: drools-core/src/main/java/org/drools/core/event/AgendaEventSupport.java [hunk 54]
```diff
@@ -39,93 +39,93 @@
 
     public AgendaEventSupport() { }
 
-    private InternalKnowledgeRuntime getKRuntime(WorkingMemory workingMemory) {
-        return ((InternalWorkingMemory) workingMemory).getKnowledgeRuntime();
+    private InternalKnowledgeRuntime getKRuntime(ReteEvaluator reteEvaluator) {
+        return reteEvaluator instanceof InternalWorkingMemory ? ((InternalWorkingMemory) reteEvaluator).getKnowledgeRuntime() : null;
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/event/AgendaEventSupport.java:43]: Can be removed? It can produce NPE

File: drools-core/src/main/java/org/drools/core/common/CompositeDefaultAgenda.java [hunk 55]
```diff
@@ -523,39 +494,19 @@ public void stageLeftTuple( RuleAgendaItem ruleAgendaItem, AgendaItem justified
...
     @Override
     public void evaluateEagerList() {
         throw new UnsupportedOperationException( "org.drools.core.common.CompositeDefaultAgenda.evaluateEagerList -> TODO" );
     }
 
     @Override
-    public Map<String, InternalActivationGroup> getActivationGroupsMap() {
-        throw new UnsupportedOperationException( "org.drools.core.common.CompositeDefaultAgenda.getActivationGroupsMap -> TODO" );
+    public void evaluateQueriesForRule(RuleAgendaItem item) {
+        throw new UnsupportedOperationException();
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/common/CompositeDefaultAgenda.java:504]: Add error message

File: drools-core/src/main/java/org/drools/core/common/AgendaGroupQueueImpl.java [hunk 56]
```diff
@@ -141,8 +141,8 @@ public SetFocusAction( String name ) {
         }
 
         @Override
-        public void execute( InternalWorkingMemory wm ) {
-            wm.getAgenda().setFocus(this.name);
+        public void execute( ReteEvaluator reteEvaluator ) {
+            ((InternalAgenda) reteEvaluator.getActivationsManager()).setFocus(this.name);
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/common/AgendaGroupQueueImpl.java:145]: Is this cast safe?

File: drools-core/src/main/java/org/drools/core/common/AgendaGroupQueueImpl.java [hunk 57]
```diff
@@ -123,8 +123,8 @@ public ClearAction( String name ) {
         }
 
         @Override
-        public void execute( InternalWorkingMemory wm ) {
-            wm.getAgenda().clearAndCancelAgendaGroup(this.name);
+        public void execute( ReteEvaluator reteEvaluator ) {
+            ((InternalAgenda) reteEvaluator.getActivationsManager()).clearAndCancelAgendaGroup(this.name);
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/common/AgendaGroupQueueImpl.java:127]: Is this cast safe?

File: drools-core/src/main/java/org/drools/core/common/ActivationsManager.java [hunk 58]
```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2010 Red Hat, Inc. and/or its affiliates.
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/common/ActivationsManager.java:2]:
```suggestion
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
```

File: drools-core/src/main/java/org/drools/core/reteoo/CompositePartitionAwareObjectSinkAdapter.java (hunk 6 above)
Comment (mariofusco) [reply drools-core/src/main/java/org/drools/core/reteoo/CompositePartitionAwareObjectSinkAdapter.java:111]: This is used only during parallel evaluation so yes it is assumed that it is `CompositeDefaultAgenda`.

File: drools-core/src/main/java/org/drools/core/reteoo/Rete.java (hunk 10 above)
Comment (mariofusco) [reply drools-core/src/main/java/org/drools/core/reteoo/Rete.java:27]: The `InternalWorkingMemory` is used for `KieSession` incremental update that won't be supported by the lightweight session.

File: drools-core/src/main/java/org/drools/core/reteoo/RightInputAdapterNode.java (hunk 12 above)
Comment (mariofusco) [reply drools-core/src/main/java/org/drools/core/reteoo/RightInputAdapterNode.java:330]: This is for incremental compilation.

File: drools-core/src/main/java/org/drools/core/common/AgendaGroupQueueImpl.java (hunk 56 above)
Comment (mariofusco) [reply drools-core/src/main/java/org/drools/core/common/AgendaGroupQueueImpl.java:145]: This is to keep supporting agenda-groups that won't be used by the lightweight session.

File: drools-core/src/main/java/org/drools/core/common/AgendaGroupQueueImpl.java (hunk 57 above)
Comment (mariofusco) [reply drools-core/src/main/java/org/drools/core/common/AgendaGroupQueueImpl.java:127]: This is to keep supporting agenda-groups that won't be used by the lightweight session.
